### ISSUE TITLE
Integration with grape-swagger: questions

### ISSUE THREAD

**Issue by nathanvda** [EmCousin/grape_fast_jsonapi#20]
Thank for this library/gem. It made my integration/setup of a json:api API using swagger very easy. 

I have two questions with regards to the integration with `grape-swagger`.

* for a single item, the swagger documentation declares each field (attribute) as a string, how can I change that. Somehow I would like to link the original class as well, so it could deduce the types of the attributes? E.g. in `PageSerializer` declare some link to the `Page` model? 
* how can we better/easier add swagger documentation for arrays/list of pages, given that I add some links and meta data to the returned result? 

Any tips or hints?

**Comment by nathanvda** [comment]
For completeness, let me add some example code, to show what I am doing, and also where I am probably diverging from the adviced path.

* i define my grape in `lib\api` 
* I have defined my serializers in `lib\api\serializers\`
* in my initializer I have the following line `GrapeSwagger.model_parsers.register(GrapeSwagger::FastJsonapi::Parser, API::Serializers::BaseSerializer)`
* all other serializers derive from `API::Serializers::BaseSerializer`

So this works, but gives a "general" documentation, it considers each attribute as string. So I derive: 

* it does find the serializer
* but somehow it does not find the model? 

So I tried adapted my `config/initializers/grape_swagger.rb` 

```
GrapeSwagger.model_parsers.register(GrapeSwagger::FastJsonapi::Parser, API::Serializers::PageSerializer)
GrapeSwagger.model_parsers.register(GrapeSwagger::FastJsonapi::Parser, API::Serializers::NewsItemSerializer)
GrapeSwagger.model_parsers.register(GrapeSwagger::FastJsonapi::Parser, API::Serializers::MenuItemSerializer)
```

The intention was: register the more "specific" serializers in the hope they are picked up correctly? I have no idea how the serializers find the active-record model for now (I am guessing it is the `set_type`? but that is just a guess). 

If I do this I get an error that `NewsItemSerializer` cannot be found, which is probably caused because I have not defined my serializers in the global namespace? However I do not understand why it is looking in global namespace, because I fully qualify every reference to the serializers. 

So my problems are

* I probably do not understand how the `GrapeSwagger.model_parsers` work? 
* I do not understand how `GrapeSwagger::FastJsonapi::Parser` finds the active record model to check the attributes against? I think it is the `set_type` function inside the serializer?

**Comment by EmCousin** [comment]
Hello @nathanvda, thank you for your issue.

> I probably do not understand how the GrapeSwagger.model_parsers work?

`GrapeSwagger::FastJsonapi::Parser` is merely a [custom parser](https://github.com/ruby-grape/grape-swagger#custom-model-parsers) that aims at providing generic documentation. Nothing actually prevents you to build your own for your specific needs for the time being. But to answer more in details your next question:

> I do not understand how GrapeSwagger::FastJsonapi::Parser finds the active record model to check the attributes against? I think it is the set_type function inside the serializer?

If you do use `ActiveRecord`, the parser will try to find the ActiveRecord model of your serializer instance by [checking its `record_type` attribute](https://github.com/EmCousin/grape_fast_jsonapi/blob/master/lib/grape_fast_jsonapi/parser.rb#L69). Its value is indeed set by the `set_type` method within your serializer.

If it finds your active record model, it then considers each attribute with the type of its related column. However if it doesn't, it fallbacks to defining each attribute as a `string` since it doesn't know how to figure its type. One possible way would be to define the attribute type like so:
```ruby
attribute: some_id, type: :integer
```
But that would have to be done within the `fast_jsonapi` gem itself.

Hope this helps

**Comment by nathanvda** [comment]
Thank you for your explanation! Just adding the type explicitly did not work for me. 

So I did a little digging, and from what I can see the parser correctly finds the serializer, and from the serializer it correctly find the model and attributes (and thus the corresponding column and types) ... so I had to dig a little deeper to see what goes wrong ...

And apparently it is on this line: 

     return map_model_attributes unless activerecord_model&.is_a?(ActiveRecord::Base)

not sure if it is specific to rails 6, because my models all derive from `ApplicationRecord` (but for me that test never seems to work, maybe it is the ruby version). I did find some more robust alternatives

*  `return map_model_attributes unless activerecord_model <= ActiveRecord::Base`
*  `return map_model_attributes unless activerecord_model <= ActiveRecord::Base`

**Comment by EmCousin** [comment]
@nathanvda It seems like https://github.com/EmCousin/grape_fast_jsonapi/pull/14 could be providing what you need. The tests are however failing and I didn't have time to test it lately. Mind to have a look possibly?

**Comment by nathanvda** [comment]
Cool! I will check it out later today 👍

**Comment by nathanvda** [comment]
Ok, I had a little trouble getting a local copy of #14 , because I had to get the work from Vincent's branch, and the test's were living in another branch (prog_rails), but I got it working. I created a new branch with the combined/fixed work. 

Or you could apply the fix for the tests, which is as easy as adding 

    require 'active_record'

at the top of the `db_record.rb` file. 

I opened a PR but will let you decide what is best. In the meantime I will close this ticket since the PR (and your help) solved all my issues.

**Comment by EmCousin** [comment]
Released version 0.2.6
